compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs

```diff
@@ -941,8 +942,24 @@ impl<'tcx> SplitWildcard<'tcx> {
                     smallvec![NonExhaustive]
                 } else if is_declared_nonexhaustive {
                     def.variants
-                        .indices()
-                        .map(|idx| Variant(idx))
+                        .iter_enumerated()
+                        .filter_map(|(idx, def)| {
+                            let x = cx.tcx.lookup_stability(def.def_id);
+                            if let Some(Stability {
+                                feature,
+                                level: StabilityLevel::Unstable { .. },
+                            }) = x
+                            {
+                                // The unstable feature is not activated so ignore this variant
+                                if !cx.tcx.stability().active_features.contains(&feature) {
+                                    None
+                                } else {
+                                    Some(Variant(idx))
+                                }
+                            } else {
+                                Some(Variant(idx))
+                            }
+                        })
```

Comment:
Ah I had missed something: right now you're filtering variants only in the `non_exhaustive` case, but we want to do it for all enums

Reply:
And I discovered https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.eval_stability which seems to be more appropriate since it checks a few more things

Reply:
I would do:
```rust
let mut ctors: SmallVec<[_; 1]> = def
    .variants
    .iter_enumerated()
    .filter(|(_, v)| {
        // Filter variants that depend on a disabled unstalbe feature.
        let is_enabled = !matches!(
            cx.tcx.eval_stability(v.def_id, None, DUMMY_SP, None),
            EvalResult::Deny { .. }
        );
        // If `exhaustive_patterns` is enabled, we exclude variants known to be
        // uninhabited.
        let is_uninhabited = cx.tcx.features().exhaustive_patterns
            && v.uninhabited_from(cx.tcx, substs, def.adt_kind(), cx.param_env)
                .contains(cx.tcx, cx.module);
        is_enabled && !is_uninhabited
    })
    .map(|(idx, _)| Variant(idx))
    .collect();
```
and then push a `NonExhaustive` into `ctors` in the `secretly_empty` and `non_exhaustive` cases.

Reply:
Perfect :+1: